src/abstractarray.jl

```diff
@@ -147,10 +147,19 @@ similar(::Type{A},::Type{T},s::Size{S}) where {A<:Array,T,S} = sizedarray_simila
 # Support tuples of mixtures of `SOneTo`s alongside the normal `Integer` and `OneTo` options
 # by simply converting them to either a tuple of Ints or a Size, re-dispatching to either one
 # of the above methods (in the case of Size) or a base fallback (in the case of Ints).
+const HeterogeneousBaseShape = Union{Integer, Base.OneTo}
 const HeterogeneousShape = Union{Integer, Base.OneTo, SOneTo}
 
-similar(A::AbstractArray, ::Type{T}, shape::Tuple{HeterogeneousShape, Vararg{HeterogeneousShape}}) where {T} = similar(A, T, homogenize_shape(shape))
-similar(::Type{A}, shape::Tuple{HeterogeneousShape, Vararg{HeterogeneousShape}}) where {A<:AbstractArray} = similar(A, homogenize_shape(shape))
+const SOneToFirstShape = Tuple{SOneTo, Vararg{HeterogeneousShape}}
+const SOneToSecondShape = Tuple{HeterogeneousBaseShape, SOneTo, Vararg{HeterogeneousShape}}
+const SOneToThirdShape = Tuple{HeterogeneousBaseShape, HeterogeneousBaseShape, SOneTo, Vararg{HeterogeneousShape}}
```

Comment:
Neater to put these into a big `Union` of all three? Then I think we can have only two extra `similar` methods rather than six.

Reply:
I thought about using a `Union` but I was just throwing the idea here first. Will simplify it.